compiler/src/iree/compiler/Codegen/Dialect/MicroKernelOps.td

```diff
@@ -0,0 +1,108 @@
+// Copyright 2023 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef IREE_CODEGEN_DIALECT_MICRO_KERNEL_OPS
+#define IREE_CODEGEN_DIALECT_MICRO_KERNEL_OPS
+
+include "iree/compiler/Codegen/Dialect/IREECodegenDialect.td"
+include "iree/compiler/Codegen/Interfaces/MicroKernelOpInterface.td"
+include "mlir/Interfaces/DestinationStyleOpInterface.td"
+
+class IREECodegen_MicroKernelOp<string mnemonic, list<Trait> traits = []> : 
+  Op<IREECodegen_Dialect, mnemonic, !listconcat(traits,
+    [DeclareOpInterfaceMethods<MicroKernelOpInterface,
+        ["lowerToFunctionCall"]>,
+     DeclareOpInterfaceMethods<DestinationStyleOpInterface>])> {}
+  
+def IREECodegen_GenericMicroKernelOp :
+    IREECodegen_MicroKernelOp<"generic_micro_kernel", [
+      AttrSizedOperandSegments]> {
+  let summary = "Generic Microkernel operator";
+
+  let description = [{
+    Operation to wrap a computation forwarded to a microkernel.
+
+    This operation is a generic representation of the DAG that is to be
+    lowered into a micro-kernel. The name of the microkernel is specified
+    as a `StrAttr`. The DAG to be forwarded is meant to be captured at
+    tensor-level. The operation implements the `DestinationStyleOpInterface`
+    so all tensors in the `outs` list must match the number and type of the
+    results of the operation.
+    After bufferization the tensor operands in `outs` are converted to
+    a memref type. At the memref-level, the operands are expected to
+    match directly into a function call with the arguments to the
+    function call being the `ins`, `outs` and `other_operands`.
+    
+    The operands of `memref` type are expected to lower to the following
+    argument sequence.
+    - `memref<f32> lowers to base_ptr.
+    - `memref<?xf32, <strides = [s0, 1], offsets = f>>
+          lowers to (base_ptr, f, index s0)
+    - `memref<?x?xf32, <strides = [s1, s0, 1], offsets = f>>
+          lowers to (base_ptr, f, s1, s0)
+    and so on....
+
+    Innermost stride of `memref`s are expected to be 1.
+    All other operands are expected to be scalar types.
+    TODO: `vector` types can be supported as well, but needs better
+    ABI specification.
+  }];
+
+  let arguments = (ins
+    StrAttr:$micro_kernel_fn_name,
+    Variadic<AnyType>:$inputs,
+    Variadic<AnyRankedTensorOrMemRefType>:$outputs,
+    Variadic<AnyType>:$other_operands);
+  let results = (outs Variadic<AnyRankedTensor>:$results);
+  let assemblyFormat = [{
+    attr-dict $micro_kernel_fn_name
+    (`ins` `(` $inputs^ `:` type($inputs) `)`)?
+    (`outs` `(` $outputs^  `:` type($outputs) `)`)?
+    (`(` $other_operands^ `:` type($other_operands) `)`)? (`->` type($results)^)?
+  }];
+}
+
+def IREECodegen_Mmt4DMicroKernelOp :
+    IREECodegen_MicroKernelOp<"mmt4d_micro_kernel", []>  {
```

Comment:
Naming choices:

- Need to settle on a consistent choice of naming for microkernels ---  `micro_kernel`, `microkernel`, `ukernel` or `uk`?   Try to be consistent also across MLIR<->C++ to make it more intuitive to navigate code? ("Why is grep not returning results for ukernel, oh it's because it's microkernel").
- Usually MLIR names have namespacing prefixes, not suffixes, and use dot (`.`) as namespace separator? E.g. `microkernel.mmt4d` or `ukernel.mmt4d` or `uk.mmt4d` ?

Reply:
+1, was wondering about the consistency of naming. I've seen that @bjacob names most of them into ukernel, maybe we can use `ukernel`?

Reply:
Naming is an Achilles heel for me. Please suggest one name, and I will go what that.

Reply:
+1 for `ukernel` everywhere.

Reply:
btw, the `.` is an IREE thing, I think

Reply:
Thanks for updating the PR, I can now see the `mmt4d.ukernel` and `generic.ukernel`.  I don't want to push the bikeshed discussion too much, but don't you think that it's more usual to have the "namespace" as a prefix rather than a suffix? Moreover, in fully qualified names like `iree_codegen.mmt4d.ukernel`, we now have a `iree_codegen` namespace as prefix and a `ukernel` namespace as suffix.  I would find `iree_codegen.ukernel.mmt4d` more natural, WDYT?

Reply:
I'm OK to do this change myself as a follow-up if you would like me to; approving this PR as-is for now, thank you so much for doing it already!

Reply:
I made the changes. Thanks!